File: eclipse/FizzBuzz/src/de/welcz/FizzBuzzEngine.java
package de.welcz;

import java.util.Arrays;
import java.util.List;

public class FizzBuzzEngine {

	private static final List<Rule> rules = Arrays.asList(
			Rule.divisibleBy(15).returns("FizzBuzz"),
			Rule.divisibleBy(3).returns("Fizz"), 
			Rule.divisibleBy(5).returns("Buzz"));

	public String calculateNext(int number) {
		return rules.stream()
				.filter(r -> r.appliesTo(number))
				.map(Rule::giveResult)
				.findFirst()
				.orElse(Integer.toString(number));
	}

}

File: ruby/FizzBuzz/src/fizz_buzz_engine.rb
require_relative 'divisible_by_extension.rb'
class FizzBuzzEngine

  def initialize
    @rules = [
        [15, 'FizzBuzz'],
        [3, 'Fizz'],
        [5, 'Buzz']
    ]
  end

  def calculate_next(number)
    matching_rule = @rules.select { |divider, _| number.divisible_by?(divider) }
                        .map { |_, result| result }.first
    matching_rule or number.to_s
  end
end